src/custom/utils/tip.test.ts

```diff
@@ -0,0 +1,268 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { FeeInformation } from '@src/custom/utils/operator'
+import {
+  ChainId,
+  Currency,
+  CurrencyAmount,
+  JSBI,
+  Pair,
+  Route,
+  Token,
+  TokenAmount,
+  Trade,
+  TradeType
+} from '@uniswap/sdk'
+import { parseUnits } from 'ethers/lib/utils'
+import { computeTradePriceBreakdown } from '../../utils/prices'
+
+// const fiveMinutesFromNow = () => new Date(Date.now() + 300000).toISOString()
+// try to parse a user entered amount for a given token
+
+// had to rebuild here as import was failing...
+function tryParseAmount(value?: string, currency?: Currency): CurrencyAmount | undefined {
+  if (!value || !currency) {
+    return undefined
+  }
+  try {
+    const typedValueParsed = parseUnits(value, currency.decimals).toString()
+    if (typedValueParsed !== '0') {
+      return currency instanceof Token
+        ? new TokenAmount(currency, JSBI.BigInt(typedValueParsed))
+        : CurrencyAmount.ether(JSBI.BigInt(typedValueParsed))
+    }
+  } catch (error) {
+    // should fail if the user specifies too many decimal places of precision (or maybe exceed max uint?)
+    console.debug(`Failed to parse input amount: "${value}"`, error)
+  }
+  // necessary for all paths to return a value
+  return undefined
+}
+
+const token1 = new Token(ChainId.MAINNET, '0x0000000000000000000000000000000000000001', 18, 'TOK_ONE')
+const token2 = new Token(ChainId.MAINNET, '0x0000000000000000000000000000000000000002', 18, 'TOK_TWO')
+const token3 = new Token(ChainId.MAINNET, '0x0000000000000000000000000000000000000003', 18, 'TOK_THREE')
+
+const pair1and2 = new Pair(
+  new TokenAmount(token1, JSBI.BigInt(100000 * 10 ** 18)),
+  new TokenAmount(token2, JSBI.BigInt(100000 * 10 ** 18))
+)
+const pair2and3 = new Pair(
+  new TokenAmount(token2, JSBI.BigInt(200000 * 10 ** 18)),
+  new TokenAmount(token3, JSBI.BigInt(300000 * 10 ** 18))
+)
+
+const DEFAULT_HOPS = { maxHops: 3, maxNumResults: 1 }
+
+interface SetupTrade {
+  type: TradeType
+  fee: Partial<FeeInformation>
+  typedAmount: string
+  inputCurrency: Currency
+  outputCurrency: Currency
+  pair: Pair
+}
+
+function setupTrade(
+  props: SetupTrade
+):
+  | (SetupTrade & {
```

Comment:
maybe u should create a type or sth for this instead of inlining the definition?

Reply:
:+1: